peekingduck/pipeline/nodes/model/fairmotv1/fairmot_files/decoder.py

```diff
@@ -0,0 +1,257 @@
+            input_shape (torch.Size): Shape of the image fed to the model.
+
+        Returns:
+            (np.ndarray): Transformed detections w.r.t. the original image
+            shape.
+        """
+        orig_h = float(orig_shape[0])
+        orig_w = float(orig_shape[1])
+        input_h = float(input_shape[2])
+        input_w = float(input_shape[3])
+
+        dets = detections.detach().cpu().numpy()
+        dets = dets.reshape(1, -1, dets.shape[2])
+        # Detection batch size has been hardcoded to 1 in FairMOT
+        return _ctdet_post_process(
+            dets[0].copy(),
+            np.array([orig_w / 2.0, orig_h / 2.0], dtype=np.float32),
+            max(input_w / input_h * orig_h, orig_w),
+            (input_w // self.down_ratio, input_h // self.down_ratio),
+        )
+
+    def _topk(
+        self, scores: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Selects top k scores and decodes to get xy coordinates.
+
+        Args:
+            scores (torch.Tensor): In the case of FairMOT, this is a heatmap
+                predicting where the object center will be.
+
+        Returns:
+            (Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor,
+            torch.Tensor]): Tuple containing top k detection scores and their
+            respective indices, classes, y-, and x- coordinates.
+        """
+        k = self.max_per_image
+        batch, cat, height, width = scores.size()
+
+        topk_scores, topk_indices = torch.topk(scores.view(batch, cat, -1), k)
+
+        topk_indices = (topk_indices % (height * width)).view(batch, -1, 1)
+        topk_y_coords = torch.div(topk_indices, width, rounding_mode="floor")
+        topk_x_coords = topk_indices % width
+
+        topk_score, topk_index = torch.topk(topk_scores.view(batch, -1), k)
+        topk_classes = torch.div(topk_index, k, rounding_mode="trunc")
+
+        topk_indices = gather_feat(topk_indices, topk_index).view(batch, k)
+        topk_y_coords = gather_feat(topk_y_coords, topk_index).view(batch, k)
+        topk_x_coords = gather_feat(topk_x_coords, topk_index).view(batch, k)
+
+        return topk_score, topk_indices, topk_classes, topk_y_coords, topk_x_coords
+
+    def _trim_outputs(self, detections: np.ndarray) -> np.ndarray:
+        """In the case of multi-class detections, trims the output to be
+        <=`max_per_image.
+
+        Args:
+            detections (np.ndarray): Object detection results.
+
+        Returns:
+            (np.ndarray): Trimmed detection results.
+        """
+        if len(detections) > self.max_per_image:  # pragma: no cover
+            # FairMOT only detects one class so this is never called.
+            kth = len(detections) - self.max_per_image
+            cut_off = np.partition(detections[:, 4], kth)[kth]
+            detections = detections[detections[:, 4] >= cut_off]
+        return detections
```

Comment:
If FairMOT only detects one class, then perhaps this function is redundant and can be removed?

Reply:
Added `# pragma: no cover` to `_trim_outputs()` and commented out the call `detections = self._trim_outputs(detections)`.

Added a comment to explain that this is not used currently, but can be potentially used if FairMOT is expanded to detect multiple classes.